### ISSUE TITLE
ParsePdf for parliament (and WHO) is using > 2GB RAM when parsing problematic PDFs

### ISSUE THREAD

**Issue by hblanks** [wellcometrust/policytool#227]
ParsePdf.who_iris is using > 2GB RAM (and exceeding its cgroup's memory limit) when parsing the PDF at:

```
s3://datalabs-staging/reach-airflow/output/policy/spider/who_iris/spider-who_iris/pdf/69/695a224a7ed725976f5c8827665cf513.pdf
```

as originally characterized in https://github.com/wellcometrust/policytool/issues/224 .

Working hypothesis: this is happening because pdf2html is producing a 128MB HTML file of gobblygook, and when we go to read/parse this, we use more RAM than we're given.

**Comment by hblanks** [comment]
OK - added logging and ran this locally against the staging bucket, and we find the memory is all happening during initial BeautifulSoup parse:

```python
        logger.info('Parsing with beautifulsoup')
        soup = bs(tf.read(), 'html.parser')
        logger.info('...parsed')

        file_pages = []
        pages = soup.find_all('page')
        full_text = soup.text

        logger.info('Iterating through pages')
        for num, page in enumerate(pages):
            words = page.find_all('text')
```

```
$ ./docker_exec.sh airflow test policy ParsePdf.who_iris 2019-09-01
/usr/local/lib/python3.6/site-packages/airflow/config_templates/airflow_local_settings.py:65: DeprecationWarning: The elasticsearch_host option in [elasticsearch] has been renamed to host - the old setting has been used, but please update your config.
  ELASTICSEARCH_HOST = conf.get('elasticsearch', 'HOST')
/usr/local/lib/python3.6/site-packages/airflow/config_templates/airflow_local_settings.py:67: DeprecationWarning: The elasticsearch_log_id_template option in [elasticsearch] has been renamed to log_id_template - the old setting has been used, but please update your config.
  ELASTICSEARCH_LOG_ID_TEMPLATE = conf.get('elasticsearch', 'LOG_ID_TEMPLATE')
/usr/local/lib/python3.6/site-packages/airflow/config_templates/airflow_local_settings.py:69: DeprecationWarning: The elasticsearch_end_of_log_mark option in [elasticsearch] has been renamed to end_of_log_mark - the old setting has been used, but please update your config.
  ELASTICSEARCH_END_OF_LOG_MARK = conf.get('elasticsearch', 'END_OF_LOG_MARK')
[2019-09-09 14:58:44,389] {settings.py:213} INFO - settings.configure_orm(): Using pool settings. pool_size=5, max_overflow=10, pool_recycle=1800, pid=285
[2019-09-09 14:58:44,768] {__init__.py:51} INFO - Using executor CeleryExecutor
[2019-09-09 14:58:45,125] {dagbag.py:90} INFO - Filling up the DagBag from /airflow/dags
[2019-09-09 14:58:45,525] {taskinstance.py:616} INFO - Dependencies all met for <TaskInstance: policy.ParsePdf.who_iris 2019-09-01T00:00:00+00:00 [None]>
[2019-09-09 14:58:45,532] {taskinstance.py:616} INFO - Dependencies all met for <TaskInstance: policy.ParsePdf.who_iris 2019-09-01T00:00:00+00:00 [None]>
[2019-09-09 14:58:45,532] {taskinstance.py:834} INFO - 
--------------------------------------------------------------------------------
[2019-09-09 14:58:45,532] {taskinstance.py:835} INFO - Starting attempt 1 of 1
[2019-09-09 14:58:45,532] {taskinstance.py:836} INFO - 
--------------------------------------------------------------------------------
[2019-09-09 14:58:45,533] {taskinstance.py:855} INFO - Executing <Task(ParsePdfOperator): ParsePdf.who_iris> on 2019-09-01T00:00:00+00:00
[2019-09-09 14:58:45,678] {main.py:244} INFO - parse_all_pdf: input_url=manifests3://datalabs-staging/reach-airflow/output/policy/spider/who_iris/spider-who_iris output_url=s3://datalabs-staging/reach-airflow/output/policy/parsed-pdfs/who_iris/parsed-pdfs-who_iris.json.gz organisation=who_iris context=2 resources_dir=None
[2019-09-09 14:58:45,696] {credentials.py:1059} INFO - Found credentials in environment variables.
[2019-09-09 14:58:45,957] {main.py:204} INFO - 695a224a7ed725976f5c8827665cf513  --- 69
[2019-09-09 14:58:54,036] {pdf_parse.py:67} INFO - Parsing with beautifulsoup
```

**Comment by hblanks** [comment]
And within 20s we're over 1.5 GB...

**Comment by jdu** [comment]
Sorry, just having a read, don't understand the whole context here but thought I would comment:

1) BeautifulSoup has been known to have some pretty terrible memory leaks on larger documents
2) `lxml` is generally faster in terms of parsing structured documents 
3) Do you actually need the intermediary step of converting to HTML and then parsing the HTML? If all you're doing is stripping text primarily, you could use a library like pikepdf to parse and iterate on the object streams in the pdfs directly removing the entire intermediary step of HTML conversion and parsing and reducing the overhead drastically.

Another alternative/tertiary option is to split the PDF and parse each individual subsection independently putting the text back together at the end, that would reduce your overall runtime overhead as well as potentially help you identify which sections of the document soup is choking on and why as you won't have to load the entire pdfs worth of HTML into soup at once.

**Comment by hblanks** [comment]
Hey! What  a treat to hear from you, @jdu . Quick notes in reply:

> 1. BeautifulSoup has been known to have some pretty terrible memory leaks on larger documents
> 2. `lxml` is generally faster in terms of parsing structured documents

Yes - indeed, my first version of the PR replaced BS with LXML for just this reason...but that broke the tests, so I figured best to fix one bug first before yanking out BeautifulSoup as well (https://github.com/wellcometrust/policytool/pull/228#issuecomment-529563854)

> 3. Do you actually need the intermediary step of converting to HTML and then parsing the HTML? If all you're doing is stripping text primarily, you could use a library like pikepdf to parse and iterate on the object streams in the pdfs directly removing the entire intermediary step of HTML conversion and parsing and reducing the overhead drastically.

We'd be delighted to spend some time on this. Before my time, @SamDepardieu was parsing PDFs directly. He ran into issues with the Python PDF library he was using, but that was long enough ago now that it's very much due for a revisit. One really nice thing about the current pipeline is at least all the PDFs we scrape are in S3 and could be churned through by an ad-hoc cluster (dask, PySpark, whatnot) with different code fairly easily.

I've tracked this as https://github.com/wellcometrust/policytool/issues/232, and also pushed a PR (to the datalabs internal repo) adding this to the product roadmap / implementation plan. (Sorry, the plan should move into this repo...it's not here yet.)
